src/compiler/checker.ts

```diff
@@ -3605,10 +3609,18 @@ module ts {
             if (!type) {
                 type = unionTypes[id] = <UnionType>createObjectType(TypeFlags.Union | getWideningFlagsOfTypes(sortedTypes));
                 type.types = sortedTypes;
+                type.reducedType = noSubtypeReduction ? undefined : type;
             }
             return type;
         }
 
+        function getReducedTypeOfUnionType(type: UnionType): Type {
+            if (!type.reducedType) {
+                type.reducedType = getUnionType(type.types);

```

Comment:
I'd explicitly pass false here, just because the purpose of this call is to explicitly reduce subtypes for a union that has already been created in the past.